### PR TITLE
handle dynamic tensor for concat operation

### PR DESCRIPTION
**Pull request by chentong319**
Signed-off-by: chentong <chentong@us.ibm.com>

Current shape inference and lowering for concat operation can not handle dynamic shape. The dynamic dimension has to be the dimension to be concatenated.

## Change in the shape inference.
Use IndexExpr to compute the output size.
## Change in lowering
Alloc with IndexExpr for output so that dynamic shape is supported.
In loop, replace integer of offset with IndexExpr. The offset is compute independently for each loop.

## issue
There are 4 concat backend tests failed in result comparison. There are 3d tensors concatenated at second and third dimension. Others passed. I compared the MLIR output of 3d_axis_0 and 3d_axis_1 and can not tell what's wrong. Suspect the tests just exposed errors after lowering.

3d_axis_0 output:
```
#map0 = affine_map<()[s0, s1] -> (s0 + s1)>
#map1 = affine_map<()[s0, s1, s2] -> (s0, s1, s2)>
#map2 = affine_map<() -> (0)>
#map3 = affine_map<() -> (2)>
#map4 = affine_map<()[s0] -> (s0)>
#map5 = affine_map<()[s0, s1, s2, s3] -> (s0 + s1, s2, s3)>
module {
  %0 = "krnl.packed_const"() {file_name = "/tmp/packed_const-5e1d66.tmp", is_le = true, size_in_bytes = 0 : i64} : () -> i64
  func @main_graph(%arg0: memref<?x2x2xf32>, %arg1: memref<?x2x2xf32>) -> memref<?x2x2xf32> attributes {input_names = ["value0", "value1"], output_names = ["output"]} {
    %c0 = constant 0 : index
    %1 = dim %arg0, %c0 : memref<?x2x2xf32>
    %2 = dim %arg1, %c0 : memref<?x2x2xf32>
    %3 = affine.apply #map0()[%1, %2]
    %4 = alloc(%3) : memref<?x2x2xf32>
    %5 = dim %arg0, %c0 : memref<?x2x2xf32>
    affine.for %arg2 = 0 to %5 {
      affine.for %arg3 = 0 to 2 {
        affine.for %arg4 = 0 to 2 {
          %7 = affine.load %arg0[symbol(%arg2), symbol(%arg3), symbol(%arg4)] : memref<?x2x2xf32>
          affine.store %7, %4[symbol(%arg2), symbol(%arg3), symbol(%arg4)] : memref<?x2x2xf32>
        }
      }
    }
    %6 = dim %arg1, %c0 : memref<?x2x2xf32>
    affine.for %arg2 = 0 to %6 {
      affine.for %arg3 = 0 to 2 {
        affine.for %arg4 = 0 to 2 {
          %7 = dim %arg0, %c0 : memref<?x2x2xf32>
          %8 = affine.load %arg1[symbol(%arg2), symbol(%arg3), symbol(%arg4)] : memref<?x2x2xf32>
          affine.store %8, %4[symbol(%arg2) + symbol(%7), symbol(%arg3), symbol(%arg4)] : memref<?x2x2xf32>
        }
      }
    }
    return %4 : memref<?x2x2xf32>
  }
  "krnl.entry_point"() {func = @main_graph, numInputs = 2 : i32, numOutputs = 1 : i32} : () -> ()
}
``` 
3d_axis_1 output:
```
#map0 = affine_map<()[s0, s1] -> (s0 + s1)>
#map1 = affine_map<()[s0, s1, s2] -> (s0, s1, s2)>
#map2 = affine_map<() -> (0)>
#map3 = affine_map<() -> (2)>
#map4 = affine_map<()[s0] -> (s0)>
#map5 = affine_map<()[s0, s1, s2, s3] -> (s0, s1 + s2, s3)>
module {
  %0 = "krnl.packed_const"() {file_name = "/tmp/packed_const-5b13cb.tmp", is_le = true, size_in_bytes = 0 : i64} : () -> i64
  func @main_graph(%arg0: memref<2x?x2xf32>, %arg1: memref<2x?x2xf32>) -> memref<2x?x2xf32> attributes {input_names = ["value0", "value1"], output_names = ["output"]} {
    %c1 = constant 1 : index
    %1 = dim %arg0, %c1 : memref<2x?x2xf32>
    %2 = dim %arg1, %c1 : memref<2x?x2xf32>
    %3 = affine.apply #map0()[%1, %2]
    %4 = alloc(%3) : memref<2x?x2xf32>
    %5 = dim %arg0, %c1 : memref<2x?x2xf32>
    affine.for %arg2 = 0 to 2 {
      affine.for %arg3 = 0 to %5 {
        affine.for %arg4 = 0 to 2 {
          %7 = affine.load %arg0[symbol(%arg2), symbol(%arg3), symbol(%arg4)] : memref<2x?x2xf32>
          affine.store %7, %4[symbol(%arg2), symbol(%arg3), symbol(%arg4)] : memref<2x?x2xf32>
        }
      }
    }
    %6 = dim %arg1, %c1 : memref<2x?x2xf32>
    affine.for %arg2 = 0 to 2 {
      affine.for %arg3 = 0 to %6 {
        affine.for %arg4 = 0 to 2 {
          %7 = dim %arg0, %c1 : memref<2x?x2xf32>
          %8 = affine.load %arg1[symbol(%arg2), symbol(%arg3), symbol(%arg4)] : memref<2x?x2xf32>
          affine.store %8, %4[symbol(%arg2), symbol(%arg3) + symbol(%7), symbol(%arg4)] : memref<2x?x2xf32>
        }
      }
    }
    return %4 : memref<2x?x2xf32>
  }
  "krnl.entry_point"() {func = @main_graph, numInputs = 2 : i32, numOutputs = 1 : i32} : () -> ()
}
```

### CODE REVIEW

File: src/Conversion/ONNXToKrnl/Tensor/Concat.cpp [hunk 1]
```diff
@@ -50,29 +52,31 @@ struct ONNXConcatOpLowering : public ConversionPattern {
       for (int r = 0; r < rank; ++r)
         inputLoops.pushBounds(0, operands[i], r);
       inputLoops.createIterateOp();
+      // writeOffset = writeOffset +
```
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToKrnl/Tensor/Concat.cpp:55]: If this is no longer needed, remove please

File: src/Dialect/ONNX/ONNXOps.cpp [hunk 2]
```diff
@@ -2162,28 +2165,23 @@ LogicalResult ONNXConcatOp::inferShapes() {
   // Negative axis means values are counted from the opposite side.
   if (axisIndex < 0) {
     axisIndex = commonRank + axisIndex;
+    // Tong Chen:
+    // TOFIX: attribute modification should be into canonicalization
```
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/ONNXOps.cpp:2169]: Oh, I like moving such changes to canonicalization... probably could do a sweep for other things as well. I would like to do the same for other constants, including tensors of constants... but I did not really know how to update them. Would like to know how if you do.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, remove comment if you have a chance; I have it then on my list of PR to merge.

File: src/Conversion/ONNXToKrnl/Tensor/Concat.cpp (hunk 1 above)
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/Concat.cpp]: deleted

File: src/Dialect/ONNX/ONNXOps.cpp (hunk 2 above)
Comment (chentong319) [reply src/Dialect/ONNX/ONNXOps.cpp:2169]: We will clean up our code later.

**Comment by AlexandreEichenberger** [comment]
@gongsu832 @caoimhinuibrian @chentong319 There seems to be a problem with the Jenkins build, 

```
docker.io/onnxmlirczar/onnx-mlir-llvmimage:s390x
+ docker build --no-cache --tag onnx-mlir-build:s390x --build-arg CPU_ARCH=s390x --build-arg BASE_IMAGE=onnxmlirczar/onnx-mlir-llvmimage:s390x -f ./docker/Dockerfile.build ./docker
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /var/lib/jenkins/workspace/ONNX-MLIR-Linux-s390x-Docker-Build/docker/Dockerfile.build: no such file or directory
Build step 'Execute shell' marked build as failure
```

Clean run is needed to merge, can one of you look into it? Tx

**Comment by gongsu832** [comment]
@AlexandreEichenberger I modified linux-s390x-docker build script for the consolidate dockerfiles pull request. This test is not essential. It's only for debugging purpose only. So you can ignore the error.

**Comment by AlexandreEichenberger** [comment]
> @AlexandreEichenberger I modified linux-s390x-docker Jenkins build script for the consolidate dockerfiles pull request. This test is not essential. It's for debugging purpose only. So you can ignore the error.

We can't, the test is required at the moment.

**Comment by gongsu832** [comment]
Yes you can. Travis does all the docker tests. As long as that passes, you are good to go. linux-s390x-docker is, as I said, for debugging purpose only.

**Pull request merged**